components/textdiff/textdiff.js

```diff
@@ -129,63 +124,70 @@ class TextDiffViewModel {
     });
   }
 
-  render(isInvalidate) {
-    return promise.resolve().then(() => {
-      if (!this.diffJson || isInvalidate) {
-        return this.getDiffJson();
-      }
-    }).then(() => {
-      if (!this.diffJson || this.diffJson.length == 0) return; // check if diffs are available (binary files do not support them)
-      let lineCount = 0;
-
-      if (!this.diffJson[0].isTrimmed) {
-        this.diffJson[0].blocks = this.diffJson[0].blocks.reduce((blocks, block) => {
-          const length = block.lines.length;
-          if (lineCount < this.loadCount) {
-            block.lines = block.lines.slice(0, this.loadCount - lineCount);
+  render() {
+    return (!this.diffJson ? this.getDiffJson() : promise.resolve())
+      .then(() => {
+        if (!this.diffJson || this.diffJson.length == 0) return; // check if diffs are available (binary files do not support them)
+
+        if (!this.diffJson[0].allBlocks) {
+          this.diffJson[0].allBlocks = this.diffJson[0].blocks;
+        }
+
+        let lineCount = 0;
+        this.diffJson[0].blocks = this.diffJson[0].allBlocks.reduce((blocks, block) => {
+          const length = (block.allLines || block.lines).length;
+          const remaining = this.loadCount - lineCount;
+          if (remaining > 0) {
+            if (remaining < length) {
+              if (!block.allLines) {
+                block.allLines = block.lines;
+              }
+              block.lines = block.allLines.slice(0, remaining);
+            } else if (block.allLines) {
+              block.lines = block.allLines;
+              block.allLines = undefined;
```

Comment:
at block level, is there a reason to have `allLines` separately?

We could loosely follow the `line count` restrictions.  if limit is 10 and we first block has 12 lines of code, we could just display all 12 lines without doing this.  

Not sure if it is worth to follow that diff line limit strictly

Reply:
It was there before (we implemented it like this 5 years ago https://github.com/FredrikNoren/ungit/commit/7bcd34c84e2516f5a54ff191879d22a10da433e7). I just changed it to enable "loading more" without invalidating the diff (reverting and really fixing https://github.com/FredrikNoren/ungit/pull/876).
But yes, beeing strict has also some bugs like this one where the whitespace changes are only visible after loading the hole block:

![image](https://user-images.githubusercontent.com/4009570/75357283-6ed30400-58b1-11ea-9e8c-b10a22269a17.png)

But one code block could also be huge like hundreds of lines and defeating the purpose of https://github.com/FredrikNoren/ungit/pull/376...

Reply:
For sure there are some tech debts here.  If I remember correctly, the main cause of inefficiencies targeted in #376 was due to animation loop moving many elements and triggering recalculations that was blocked by animation loop due to large diff box opened up, which ultimately addressed in #630.  Performance part of the incremental load shouldn't be a consideration any more. 

However, if we are concerned by unexpectedly adding large block.  We could do conservative approach and load only that is below the limit.

Reply:
I have changed the logic to always load complete blocks as long as the load limit is not reached.